### DISCUSSION TITLE
Can it be used as library or Framework ?

### DISCUSSION THREAD

**Discussion by aryan-more** [stalwartlabs/smtp-server#10] [Q&A] [answered]
So, I wanted to create a mail server with my own custom features, such as:

1) Allowing me to receive mail with any username (excluding the domain part), like a@example.com or b@example.com and forwarding it to my LMTP server.
2) Allowing me to send mail from any username (excluding the domain part), but with custom authentication logic for security reasons.

For context, I will be using this mail server solely for myself, so there is no need for user management. Additionally, I want to be able to use email addresses on-the-fly, such as using x@example.com when signing up for a website.

I tried to implement this using Golang with go-smtp package from https://github.com/emersion/go-smtp, but I got stuck on the sending part as the package doesn't support it.

So I'm bit confused how can  i achieve , but I'm thinking about creating my own fork for such thing so what should i do?

**Comment by mdecimus** [comment, marked as answer]
> Allowing me to receive mail with any username (excluding the domain part), like [a@example.com](mailto:a@example.com) or [b@example.com](mailto:b@example.com) and forwarding it to my LMTP server.

This can be done with Stalwart SMTP by tweaking the configuration files.

> Allowing me to send mail from any username (excluding the domain part), but with custom authentication logic for security reasons.

This could also be done from the config file by using custom rules and/or Sieve scripts. It will depend on what kind of custom authentication logic you had in mind.

> So I'm bit confused how can i achieve , but I'm thinking about creating my own fork for such thing so what should i do?

Yes, you could either fork the entire SMTP server, of you could also write a simple SMTP server by using the [smtp-proto](https://github.com/stalwartlabs/smtp-proto) and [mail-send](https://github.com/stalwartlabs/mail-send) crates.

**Reply by aryan-more** [reply]
I want to make api call for verification using PLAIN AUTH ,but i don't want to use imap4

**Comment by i18nsite** [comment]
mail-send can I use this ，and  find the peer server by mx and directly connect to the peer server to send emails?

**Reply by mdecimus** [reply]
Only mail-send can be used as a library, but not the smtp-server currently.
